### PR TITLE
add coherency function

### PR DESCRIPTION
**Pull request by leoscholl**
work in progress

### CODE REVIEW

File: aopy/analysis/base.py [hunk 1]
```diff
@@ -1227,6 +1226,187 @@ def interp_nans(x):
...
+
+        # tmp is shape (N, nch, ntr)
+        # tapers is shape (N, k)
+
+        # Compute power for channel 1
+        ch1_rolled = np.reshape(tapers[:,:,np.newaxis]*tmp[:,[ch1],:], (win_size, k*ntr))
+        Xk_tmp1 = scipy.fft.fft(ch1_rolled.T, nf, axis=-1, overwrite_x=True, workers=workers).T
+        Xk1 = Xk_tmp1[int(nfk[0]):int(nfk[1])]
+        S1 = np.sum(Xk1*Xk1.conj(), axis=1) # sum across tapers and trials
+
```
Comment (resatomo) [inline aopy/analysis/base.py:1390]: Should we take root? because it is now squared power.

**Comment by leoscholl** [comment]
It’s just convenient to not have to square it later. I can add some commentsOn Sep 13, 2023, at 6:44 PM, Tomo ***@***.***> wrote:﻿
@resatomo commented on this pull request.



In aopy/analysis/base.py:
> +    for win in range(nwin):
+        if ref:
+            mX = np.mean(data[dn*win:dn*win+win_size], axis=1, keepdims=True) # Mean across channels for that window
+            tmp = (data[dn*win:dn*win+win_size]-mX[:,0,...]) # Subtract mean from data 
+        else:
+            tmp = (data[dn*win:dn*win+win_size])
+
+        # tmp is shape (N, nch, ntr)
+        # tapers is shape (N, k)
+
+        # Compute power for channel 1
+        ch1_rolled = np.reshape(tapers[:,:,np.newaxis]*tmp[:,[ch1],:], (win_size, k*ntr))
+        Xk_tmp1 = scipy.fft.fft(ch1_rolled.T, nf, axis=-1, overwrite_x=True, workers=workers).T
+        Xk1 = Xk_tmp1[int(nfk[0]):int(nfk[1])]
+        S1 = np.sum(Xk1*Xk1.conj(), axis=1) # sum across tapers and trials
+

Should we take root? because it is now squared power.

File: aopy/analysis/base.py [hunk 2]
```diff
@@ -1227,6 +1226,187 @@ def interp_nans(x):
...
+        S12 = np.sum(Xk*Yk.conj(), axis=1) # sum across tapers and trials
+    
+        # Coherence
+        coh = np.abs(S12/np.sqrt(S1*S2))**2
+
+        # Imaginary coherence
+        coh = np.abs(np.imag(S12/np.sqrt(S1*S2)))
+
+    Args:
+        data ((nt,nch,ntr) array): evoked potential across all channels and trials
```
Comment (resatomo) [inline aopy/analysis/base.py:1251]: ch argument

**Review by resatomo — APPROVED** [review]
looks good

**Pull request merged**